src/core/hrs.ml

```diff
@@ -35,39 +35,50 @@ let print_term : bool -> term pp = fun lhs ->
     | Appl(t,u)    -> out "app(%a,%a)" pp t pp u
     | Abst(a,t)    ->
         let (x, t) = Bindlib.unbind t in
-        if lhs then out "lam(m_typ,\\v_%s.%a)" (Bindlib.name_of x) pp t
-        else out "lam(%a,\\v_%s.%a)" pp a (Bindlib.name_of x) pp t
+        if lhs then out "lam(m_typ,\\%s.%a)" (Bindlib.name_of x) pp t
+        else out "lam(%a,\\%s.%a)" pp a (Bindlib.name_of x) pp t
     | Prod(a,b)    ->
         let (x, b) = Bindlib.unbind b in
-        out "pi(%a,\\v_%s.%a)" pp a (Bindlib.name_of x) pp b
+        out "pi(%a,\\%s.%a)" pp a (Bindlib.name_of x) pp b
   in pp
 
-(** [print_rule oc s r] outputs the rule declaration corresponding [r] (on the
-    symbol [s]), to the output channel [oc]. *)
-let print_rule : Format.formatter -> sym -> rule -> unit = fun oc s r ->
-  (* Print the pattern variable declarations. *)
-  let get_patt_names : term list -> string list = fun ts ->
-    let names = Stdlib.ref [] in
+(** [print_rule_aux oc lhs rhs] outputs the rule declaration [lhs]->[rhs]
+    to the output channel [oc] *)
+let print_rule_aux : Format.formatter -> term -> term -> unit =
+  fun oc lhs rhs ->
+  (* gets pattern and binded variable names *)
+  let add_var_names : StrSet.t -> term -> StrSet.t = fun ns t ->
+    let names = Stdlib.ref ns in
     let fn t =
       match t with
-      | Patt(_,n,_) -> Stdlib.(names := n :: !names)
+      | Patt(_,n,_) -> Stdlib.(names := StrSet.add ("&" ^ n) !names)
+      | Abst(_,b) ->
+        let (x, _) = Bindlib.unbind b in
+        Stdlib.(names := StrSet.add (Bindlib.name_of x) !names)
       | _           -> ()
     in
-    List.iter (Basics.iter fn) ts;
-    List.sort_uniq String.compare Stdlib.(!names)
+    Basics.iter fn t;
+    Stdlib.(!names)
   in
-  let names = get_patt_names r.lhs in
-  if names <> [] then
+  let names = add_var_names StrSet.empty lhs in
+  let names = add_var_names names rhs in
+  if not (StrSet.is_empty names) then
     begin
-      let print_name oc x = Format.fprintf oc "  &%s : term" x in
+      let print_name oc x = Format.fprintf oc "  %s : term" x in
+      let names = StrSet.elements names in
       Format.fprintf oc "(VAR\n%a\n)\n" (List.pp print_name "\n") names
     end;
   (* Print the rewriting rule. *)
-  let lhs = Basics.add_args (Symb(s,Qualified)) r.lhs in
   Format.fprintf oc "(RULES %a" (print_term true) lhs;
-  let rhs = Basics.term_of_rhs r in
   Format.fprintf oc "\n    -> %a)\n" (print_term false) rhs
 
+(** [print_rule oc s r] outputs the rule declaration corresponding [r] (on the
+    symbol [s]), to the output channel [oc]. *)
+let print_rule : Format.formatter -> sym -> rule -> unit = fun oc s r ->
```

Comment:
I would prefer to rename this function into print_sym_rule, and the function print_rule_aux into print_rule.